lib/block_view_lockups.go

```diff
@@ -0,0 +1,2178 @@
+		return 0, 0, nil, errors.Wrapf(RuleErrorCoinUnlockNoUnlockableCoinsFound,
+			"_connectCoinUnlock")
+	}
+
+	// Unlock coins until the amount specified by the transaction is deducted.
+	var prevLockedBalanceEntries []*LockedBalanceEntry
+	unlockedBalance := uint256.NewInt()
+	for _, unlockableLockedBalanceEntry := range unlockableLockedBalanceEntries {
+		unlockedBalance, err =
+			SafeUint256().Add(unlockedBalance, &unlockableLockedBalanceEntry.BalanceBaseUnits)
+		if err != nil {
+			return 0, 0, nil,
+				errors.Wrapf(RuleErrorCoinUnlockUnlockableCoinsOverflow, "_connectCoinUnlock")
+		}
+
+		// Append the LockedBalanceEntry in the event we rollback the transaction.
+		prevLockedBalanceEntries = append(prevLockedBalanceEntries, unlockableLockedBalanceEntry.Copy())
+
+		// Update the LockedBalanceEntry and delete the record.
+		unlockableLockedBalanceEntry.BalanceBaseUnits = *uint256.NewInt()
+		bav._deleteLockedBalanceEntry(unlockableLockedBalanceEntry)
+	}
+
+	// Credit the transactor with either DAO coins or DeSo for this unlock.
+	var prevTransactorBalanceEntry *BalanceEntry
+	var prevCoinEntry *CoinEntry
+	if profilePKID.IsZeroPKID() {
+		// Ensure the uint256 can be properly represented as a uint64.
+		if !unlockedBalance.IsUint64() {
+			return 0, 0, nil,
+				errors.Wrapf(RuleErrorCoinUnlockUnlockableDeSoOverflow, "_connectCoinUnlock")
+		}
+
+		// Add the unlockedBalance to the transactors DeSo balance.
+		// NOTE: _addBalance checks for balance overflow.
+		utxoOp, err := bav._addBalance(unlockedBalance.Uint64(), txn.PublicKey)
+		if err != nil {
+			return 0, 0, nil, errors.Wrapf(err, "_connectCoinUnlock: error"+
+				"adding CoinToUnlockBaseUnits to the transactor balance: ")
+		}
+		utxoOpsForTxn = append(utxoOpsForTxn, utxoOp)
+	} else {
+		prevTransactorBalanceEntry = bav._getBalanceEntryForHODLerPKIDAndCreatorPKID(hodlerPKID, profilePKID, true)
+
+		// Credit the transactor with the unlock amount.
+		newTransactorBalanceEntry := prevTransactorBalanceEntry.Copy()
+		newTransactorBalanceNanos, err := SafeUint256().Add(&newTransactorBalanceEntry.BalanceNanos, unlockedBalance)
+		if err != nil {
+			return 0, 0, nil, errors.Wrapf(RuleErrorCoinUnlockCausesBalanceOverflow,
+				"_connectCoinUnlock")
+		}
+		newTransactorBalanceEntry.BalanceNanos = *newTransactorBalanceNanos
+		bav._setBalanceEntryMappings(newTransactorBalanceEntry, true)
+
+		// Update CoinsInCirculation and NumberOfHolders to accurately reflect the changing balance.
+		prevCoinEntry = profileEntry.DAOCoinEntry.Copy()
+		newCoinsInCirculationNanos, err := SafeUint256().Add(
+			&profileEntry.DAOCoinEntry.CoinsInCirculationNanos,
+			unlockedBalance)
+		if err != nil {
+			return 0, 0, nil, errors.Wrapf(RuleErrorCoinUnlockCausesCoinsInCirculationOverflow,
+				"_connectCoinUnlock")
+		}
+		profileEntry.DAOCoinEntry.CoinsInCirculationNanos = *newCoinsInCirculationNanos
+		if prevTransactorBalanceEntry.BalanceNanos.IsZero() {
```

Comment:
Is it possible that the previous balance was zero but you ended up unlocking nothing and so you increment this erroneously? If you just check that the AFTER balance is >0 in this if you're good.